csharp/test/Drivers/Interop/Snowflake/CastTests.cs

```diff
@@ -0,0 +1,440 @@
+        [InlineData(TIMESTAMP_TZ, "'1970-01-01 12:34:56+0000'", "1970-01-01 00:00:00+0000", ArrowTypeId.Date32, TO_DATE)] // Date portion, only
+        [InlineData(TIMESTAMP_TZ, "'2970-01-01 12:00:00+0000'", "2970-01-01 12:00:00.000 Z", ArrowTypeId.String, TO_VARCHAR)]
+        [InlineData(TIMESTAMP_TZ, "'2970-01-01 12:00:00+0000'", "Jan 01, 2970", ArrowTypeId.String, TO_VARCHAR, COLUMN_NAME + ", 'mon dd, yyyy'")]
+#if NET6_0_OR_GREATER
+        [InlineData(TIMESTAMP_TZ, "'2970-01-01 12:00:00+0000'", "1970-01-01 12:00:00+0000", ArrowTypeId.Time64, TO_TIME)] // Time portion, only
+#else
+        [InlineData(TIMESTAMP_TZ, "'2970-01-01 12:00:00+0000'", 43200000000, ArrowTypeId.Time64, TO_TIME)] // Microseconds
+#endif
+        public void TestCastPositive(
+            string columnSpecification,
+            string sourceValue,
+            object expectedValue,
+            ArrowTypeId expectedType,
+            string castFunction,
+            string castExpression = null)
+        {
+            InitializeTest(columnSpecification, sourceValue, out string columnName, out string table);
+            SelectWithCastAndValidateValue(
+                table,
+                castFunction,
+                castExpression ?? columnName,
+                expectedValue,
+                expectedType);
+        }
+
+        [SkippableTheory]
+        [InlineData(NUMERIC, "2345", typeof(AdbcException), TO_VARCHAR, COLUMN_NAME + ", 123", new[] { "42601", "SQL compilation error" })] // Invalid format type.
+        [InlineData(NUMERIC, "2345", typeof(AdbcException), TO_VARCHAR, COLUMN_NAME + ", '123'", new[] { "22007", "Bad output format" })] // Invalid format type.
+        [InlineData(VARCHAR, "'ABC'", typeof(AdbcException), TO_NUMERIC, null, new[] { "22018", "Numeric value" })] // Non numeric value
+        [InlineData(VARCHAR, "'3'", typeof(AdbcException), TO_BOOLEAN, null, new[] { "22018", "Boolean value" })] // Non boolean value
+        [InlineData(VARCHAR, "'31/12/1970'", typeof(AdbcException), TO_TIMESTAMP_TZ, null, new[] { "22007", "Timestamp" })] // Non date value
+        public void TestCastNegative(
+            string columnSpecification,
+            string sourceValue,
+            Type expectedExceptionType,
+            string castFunction,
+            string castExpression = null,
+            string[] expectedExceptionTextContains = null)
+        {
+            InitializeTest(columnSpecification, sourceValue, out string columnName, out string table);
+            SelectWithCastAndValidateException(
+                table,
+                castFunction,
+                castExpression ?? columnName,
+                expectedExceptionType,
+                expectedExceptionTextContains);
+        }
+
+        [SkippableTheory]
+        [InlineData(ARRAY, "SELECT ARRAY_CONSTRUCT('TRUE', 'FALSE')", "[\n  \"TRUE\",\n  \"FALSE\"\n]", ArrowTypeId.String, TO_ARRAY)]
```

Comment:
Note: Is it acceptable to expect that ARRAY/OBJECT/VARIANT all return String data representation? Or should it deserialize in a native data type?

Reply:
Until Arrow itself has better support for VARIANT types, I don't know that we can realistically do much better than this.